File: backend/app/schemas/practice.py
"""
Pydantic schemas for Practice Mode
"""
from pydantic import BaseModel, Field
from typing import List, Optional
from datetime import datetime
from enum import Enum


class DifficultyLevel(str, Enum):
    EASY = "easy"
    MEDIUM = "medium"
    HARD = "hard"


class SessionStatus(str, Enum):
    IN_PROGRESS = "in_progress"
    COMPLETED = "completed"
    ABANDONED = "abandoned"


# ============================================================================
# Curriculum Topics
# ============================================================================

class TopicInfo(BaseModel):
    """Topic information for selection UI"""
    id: str
    subject: str
    topic: str
    display_name: str
    description: Optional[str] = None
    subtopics: List[str] = []
    icon: Optional[str] = None
    question_count: int = 0


class TopicsResponse(BaseModel):
    """Response with available topics"""
    class_level: int
    subjects: List[str]
    topics: List[TopicInfo]


# ============================================================================
# Practice Session
# ============================================================================

class StartSessionRequest(BaseModel):
    """Request to start a new practice session"""
    subject: str
    topic: Optional[str] = None  # None = mixed topics
    difficulty: Optional[DifficultyLevel] = None  # None = adaptive
    question_count: int = Field(default=10, ge=5, le=30)
    time_limit_seconds: Optional[int] = Field(default=None, ge=60)  # None = no limit


class StartSessionResponse(BaseModel):
    """Response after starting a session"""
    session_id: str
    subject: str
    topic: Optional[str]
    difficulty: Optional[str]
    question_count: int
    time_limit_seconds: Optional[int]
    started_at: datetime


# ============================================================================
# Questions
# ============================================================================

class QuestionForSession(BaseModel):
    """Question delivered during a session (no answer revealed)"""
    id: str
    question_order: int
    question: str
    options: List[str]
    subject: str
    topic: str
    difficulty: str
    time_estimate_seconds: int


class NextQuestionResponse(BaseModel):
    """Response with the next question"""
    session_id: str
    question: QuestionForSession
    current_question_number: int
    total_questions: int
    time_remaining_seconds: Optional[int]
    session_elapsed_seconds: int


# ============================================================================
# Answer Submission
# ============================================================================

class SubmitAnswerRequest(BaseModel):
    """Submit answer for current question"""
    answer: str
    time_taken_seconds: int = Field(ge=0)


class SubmitAnswerResponse(BaseModel):
    """Response after submitting an answer"""
    is_correct: bool
    correct_answer: str
    explanation: str
    time_taken_seconds: int
    points_earned: int = 0  # For gamification later

    # Progress
    questions_answered: int
    questions_remaining: int
    current_score: int  # Correct answers so far
    current_accuracy: float


# ============================================================================
# Session End & Review
# ============================================================================

class QuestionReview(BaseModel):
    """Question with user's answer for review"""
    question_order: int
    question: str
    options: List[str]
    correct_answer: str
    user_answer: Optional[str]
    is_correct: Optional[bool]
    explanation: str
    time_taken_seconds: Optional[int]
    subject: str
    topic: str
    difficulty: str


class SessionSummary(BaseModel):
    """Summary of completed session"""
    session_id: str
    status: SessionStatus
    subject: str
    topic: Optional[str]

    # Score
    total_questions: int
    correct_answers: int
    wrong_answers: int
    skipped: int
    score_percentage: float

    # Time
    total_time_seconds: int
    avg_time_per_question: float
    time_limit_seconds: Optional[int]

    # Breakdown by difficulty
    easy_correct: int = 0
    easy_total: int = 0
    medium_correct: int = 0
    medium_total: int = 0
    hard_correct: int = 0
    hard_total: int = 0

    # Weak areas identified
    weak_topics: List[str] = []
    strong_topics: List[str] = []

    started_at: datetime
    ended_at: datetime


class EndSessionRequest(BaseModel):
    """Request to end session (early or normal)"""
    reason: Optional[str] = None  # For abandoned sessions


class EndSessionResponse(BaseModel):
    """Response after ending session"""
    summary: SessionSummary
    questions_review: List[QuestionReview]


# ============================================================================
# Concept Scores / Progress
# ============================================================================

class ConceptMastery(BaseModel):
    """Mastery info for a single concept/topic"""
    subject: str
    topic: str
    subtopic: Optional[str] = None
    display_name: str

    mastery_score: float  # 0-100
    total_attempts: int
    correct_attempts: int
    accuracy: float

    current_streak: int
    best_streak: int

    recommended_difficulty: DifficultyLevel
    last_practiced_at: Optional[datetime]


class UserProgress(BaseModel):
    """User's overall progress across concepts"""
    user_id: str
    class_level: int

    # Overall stats
    total_sessions: int
    total_questions_attempted: int
    overall_accuracy: float
    total_study_time_minutes: int

    # By subject
    subject_scores: dict  # {subject: accuracy}

    # Concept mastery list
    concepts: List[ConceptMastery]

    # Recommendations
    weak_areas: List[str]
    suggested_topics: List[TopicInfo]


# ============================================================================
# Session History
# ============================================================================

class SessionHistoryItem(BaseModel):
    """Summary of a past session for history list"""
    session_id: str
    subject: str
    topic: Optional[str]
    score_percentage: float
    total_questions: int
    correct_answers: int
    total_time_seconds: int
    started_at: datetime
    status: SessionStatus


class SessionHistoryResponse(BaseModel):
    """Paginated session history"""
    sessions: List[SessionHistoryItem]
    total_count: int
    page: int
    page_size: int

File: web/src/api/practice.ts
import { apiClient } from './client';

/**
 * Practice API Service
 * 
 * Handles all practice-related API calls
 */

export interface TopicInfo {
  id: string;
  subject: string;
  topic: string;
  display_name: string;
  description?: string | null;
  subtopics: string[];
  icon?: string | null;
  question_count: number;
}

export interface TopicsResponse {
  class_level: number;
  subjects: string[];
  topics: TopicInfo[];
}

export type DifficultyLevel = 'easy' | 'medium' | 'hard';

export interface StartSessionRequest {
  subject: string;
  topic?: string | null;
  difficulty?: DifficultyLevel | null;
  question_count?: number;
  time_limit_seconds?: number | null;
}

export interface StartSessionResponse {
  session_id: string;
  subject: string;
  topic: string | null;
  difficulty: string | null;
  question_count: number;
  time_limit_seconds: number | null;
  started_at: string;
}

export interface QuestionForSession {
  id: string;
  question_order: number;
  question: string;
  options: string[];
  subject: string;
  topic: string;
  difficulty: string;
  time_estimate_seconds: number;
}

export interface NextQuestionResponse {
  session_id: string;
  question: QuestionForSession;
  current_question_number: number;
  total_questions: number;
  time_remaining_seconds: number | null;
  session_elapsed_seconds: number;
}

export interface SubmitAnswerRequest {
  answer: string;
  time_taken_seconds: number;
}

export interface SubmitAnswerResponse {
  is_correct: boolean;
  correct_answer: string;
  explanation: string;
  time_taken_seconds: number;
  points_earned: number;
  questions_answered: number;
  questions_remaining: number;
  current_score: number;
  current_accuracy: number;
}

export interface QuestionReview {
  question_order: number;
  question: string;
  options: string[];
  correct_answer: string;
  user_answer: string | null;
  is_correct: boolean | null;
  explanation: string;
  time_taken_seconds: number | null;
  subject: string;
  topic: string;
  difficulty: string;
}

export interface SessionSummary {
  session_id: string;
  status: 'in_progress' | 'completed' | 'abandoned';
  subject: string;
  topic: string | null;
  total_questions: number;
  correct_answers: number;
  wrong_answers: number;
  skipped: number;
  score_percentage: number;
  total_time_seconds: number;
  avg_time_per_question: number;
  time_limit_seconds: number | null;
  easy_correct: number;
  easy_total: number;
  medium_correct: number;
  medium_total: number;
  hard_correct: number;
  hard_total: number;
  weak_topics: string[];
  strong_topics: string[];
  started_at: string;
  ended_at: string;
}

export interface EndSessionResponse {
  summary: SessionSummary;
  questions_review: QuestionReview[];
}

export interface SessionHistoryItem {
  session_id: string;
  subject: string;
  topic: string | null;
  score_percentage: number;
  total_questions: number;
  correct_answers: number;
  total_time_seconds: number;
  started_at: string;
  status: 'in_progress' | 'completed' | 'abandoned';
}

export interface SessionHistoryResponse {
  sessions: SessionHistoryItem[];
  total_count: number;
  page: number;
  page_size: number;
}

/**
 * Get available topics for practice
 */
export async function getTopics(subject?: string): Promise<TopicsResponse> {
  const params = subject ? { subject } : {};
  const response = await apiClient.get<TopicsResponse>('/practice/topics', { params });
  return response.data;
}

/**
 * Start a new practice session
 */
export async function startSession(request: StartSessionRequest): Promise<StartSessionResponse> {
  const response = await apiClient.post<StartSessionResponse>('/practice/session/start', request);
  return response.data;
}

/**
 * Get the next question in a session
 */
export async function getNextQuestion(sessionId: string): Promise<NextQuestionResponse> {
  const response = await apiClient.get<NextQuestionResponse>(`/practice/session/${sessionId}/next`);
  return response.data;
}

/**
 * Submit an answer for the current question
 */
export async function submitAnswer(
  sessionId: string,
  request: SubmitAnswerRequest
): Promise<SubmitAnswerResponse> {
  const response = await apiClient.post<SubmitAnswerResponse>(
    `/practice/session/${sessionId}/submit`,
    request
  );
  return response.data;
}

/**
 * End a practice session
 */
export async function endSession(sessionId: string, reason?: string): Promise<EndSessionResponse> {
  const response = await apiClient.post<EndSessionResponse>(
    `/practice/session/${sessionId}/end`,
    reason ? { reason } : {}
  );
  return response.data;
}

/**
 * Get session review (for completed sessions)
 */
export async function getSessionReview(sessionId: string): Promise<EndSessionResponse> {
  const response = await apiClient.get<EndSessionResponse>(`/practice/session/${sessionId}/review`);
  return response.data;
}

/**
 * Get session history
 */
export async function getSessionHistory(
  page: number = 1,
  pageSize: number = 10
): Promise<SessionHistoryResponse> {
  const response = await apiClient.get<SessionHistoryResponse>('/practice/history', {
    params: { page, page_size: pageSize },
  });
  return response.data;
}

